### DISCUSSION TITLE
Swayidle: enable/disable monitor

### DISCUSSION THREAD

**Discussion by manueldiagostino** [hyprwm/Hyprland#1051] [Q&A] [answered]
Hi,
I would use [swayidle](https://github.com/swaywm/swayidle) to manage the automatic locking of the system. The command is something like 

```
swayidle -w \
        timeout 300 '~/.config/hypr/lock.sh' \
        timeout 360 'swaymsg "output * dpms off"' \
        resume 'swaymsg "output * dpms on"' \
        before-sleep '~/.config/hypr/lock.sh' \
        lock '~/.config/hypr/lock.sh'
```

where _lock.sh_ is a script that uses [swaylock-effects](https://github.com/mortie/swaylock-effects).
Is it correct using `hyprctl dispatch monitor name,enable|disable` in place of `swaymsg "output * dpms on|off"`,  to turn the monitor on/off?

**Comment by vaxerski** [comment, marked as answer]
no. dpms is `hyprctl dispatch dpms off` (on for enable)

**Reply by manueldiagostino** [reply]
Thanks!

**Comment by abimagnus** [comment]
Hello, @manueldiagostino is this working fine? for me its not..below is my command in hyprland.conf

```exec-once=swayidle -w timeout 300 'swaylock' timeout 360 'hyprctl dispatch dpms off' resume 'hyprctl dispatch dpms on' before-sleep 'swaylock -f'```

Monitor is not getting off after locked..if i enter ```hyprctl dispatch dpms off``` in terminal its working fine but not with swayidle..

**Reply by manueldiagostino** [reply]
Hi @abimagnus, from `man swayidle(1)`:

> To make sure swayidle waits for swaylock to lock the screen before it releases 
> the inhibition lock, the -w options is used in swayidle, and -f in swaylock.

so make sure to use `-f, --daemonize` option when you call swaylock, maybe it can helps.
If you want, check my [config](https://github.com/manueldiagostino/wayland_dotfiles).

**Comment by abimagnus** [comment]
My bad, that works perfectly Thank you very much @manueldiagostino and thank you @vaxerski  for this amazing work ! 👍🏼

**Comment by manu0600** [comment]
I would like to report a bug related to this topic: 

I have the following line in my config:
`exec-once = swayidle -w timeout 180 'swaylock -f' timeout 360 'hyprctl dispatch dpms off' resume 'hyprctl dispatch dpms on' before-sleep 'swaylock -f'`
and it works well when I use it on my laptop, but when I plug an HDMI monitor too, there is something strange happening: after 3min it locks correctly, then 3min later the displays turn off (both laptop and external), but then about 10 seconds after that, the displays turn back on and I could unlock if needed, but if I continue to leave it be, it will wait another 6 min, turn off both screens, wait 10sec and wake them up. I feel like this is going to continue forever, BUT after 2-3 attempts like that, swaylock crashes and gives a red screen. From there I have to force shutdown the machine and lose all my ongoing work

Do you know how I could troubleshoot that?

**Reply by manueldiagostino** [reply]
Have you tried calling it for each single monitor?
 ``` bash
hyprctl dispatch dpms on/off <hdmi_monitor_name> && hyprctl dispatch dpms on/off <main_monitor_name>
```

**Reply by manu0600** [reply]
Actually I tested a few things: it turns out when I simply run in a terminal `hyprctl dispatch dpms off HDMI-A-1`, it turns off, and then 10sec later it turns back on. So it's not related to swayidle, but do you know what could it be?

**Reply by manueldiagostino** [reply]
I don't know, I use the same command [here](https://github.com/manueldiagostino/wayland_dotfiles/blob/main/.config/hypr/idle.sh) without problems.

**Reply by Atrate** [reply]
> Actually I tested a few things: it turns out when I simply run in a terminal `hyprctl dispatch dpms off HDMI-A-1`, it turns off, and then 10sec later it turns back on. So it's not related to swayidle, but do you know what could it be?

I can confirm that the same thing happens to me, red screen crash included.

**Comment by krakow10** [comment]
I would like to report a bug related to this topic:

Using hypridle or swayidle (both have the issue), the monitors can go to sleep successfully using dpms, but 2/3 do not wake up properly.  My Asus 165Hz 1440p monitor wakes normally, but the LG 4K monitors flanking it do not.  What happens is that the side monitors stay black unless I am continuously moving the mouse (on any monitor), at which point they wake, but if I stop moving the mouse they go back to sleep.

Reloading configuration or switching dpms off and on again does not fix the issue.

Changing the configuration to disable the side monitors:
```
monitor=DP-1, disable
monitor=DP-3, disable
```

And then back to non-disabled hyprland.conf does fix the issue completely, but is annoying to do every time my monitors sleep.

I am using the following hypridle configuration:
```
general {
    lock_cmd = notify-send "lock!"          # dbus/sysd lock command (loginctl lock-session)
    unlock_cmd = notify-send "unlock!"      # same as above, but unlock
    before_sleep_cmd = notify-send "Zzz"    # command ran before sleep
    after_sleep_cmd = notify-send "Awake!"  # command ran after sleep
    ignore_dbus_inhibit = false             # whether to ignore dbus-sent idle-inhibit requests (used by e.g. firefox or steam)
    ignore_systemd_inhibit = false          # whether to ignore systemd-inhibit --what=idle inhibitors
}

listener {
    timeout = 600                           # in seconds
    on-timeout = hyprctl dispatch dpms off  # command to run when timeout has passed
    on-resume = hyprctl dispatch dpms on    # command to run when activity is detected after timeout has fired.
}
```

**Reply by krakow10** [reply]
I made a script to fix it using a keybind if anyone is having a similar issue:
`fix-hyprland.sh`
```bash
hyprctl keyword monitor DP-1, disable
hyprctl keyword monitor DP-3, disable
sleep 1
hyprctl keyword monitor DP-1, 3840x2160@60, 0x0, 2
hyprctl keyword monitor DP-3, 3840x2160@60, 3200x0, 2
```
and then in `hyprland.conf`:
```
bind = $mainMod, U, exec, sh /home/quat/fix-hyprland.sh
```
